staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -105,8 +107,14 @@ func (w *watcher) Watch(ctx context.Context, key string, rev int64, opts storage
 	if opts.ProgressNotify && w.newFunc == nil {
 		return nil, apierrors.NewInternalError(errors.New("progressNotify for watch is unsupported by the etcd storage because no newFunc was provided"))
 	}
-	wc := w.createWatchChan(ctx, key, rev, opts.Recursive, opts.ProgressNotify, opts.Predicate)
-	go wc.run()
+	initialEventsEndBookmarkRequired := w.isInitialEventsEndBookmarkRequired(opts)
+	startWatchRV, err := w.getStartWatchResourceVersion(ctx, rev, opts)
+	if err != nil {
+		return nil, err
+	}
+	forceInitialEvents := w.areInitialEventsRequired(startWatchRV, opts)
```

Comment:
I can inline, np.

why not to pass `startWatchRV` to the `areInitialEventsRequired` method?

Reply:
For me, it makes it hard to reason about, as startWatchRV is not the value passed in the request, but rather something that we recomputed (and the logic is unclear).
It's easy to say what I want if I see the "user input". It's no longer that easy if I see f(user input), even if I know f

Understanding if that's exactly what we want in all case just requires much more thinking [I think it's still correct, but it's much harder to reason about]

Reply:
For me it is less intuitive since we overwrite `rev` with  `startWatchRV`. This also affects `areInitialEventsRequired` because of the legacy case (`rev=0`) we have there. 

Nevertheless I can slightly change the legacy case to match your expectations. But then the logic would be even less intuitive since it would be scattered between `getStartWatchResourceVersion` and  `areInitialEventsRequired`.

Reply:
I don't agree.
We're computing whether certain actions (listing initial events or sending bookmark) should be taken. I want a simple logic that tells me true/false based on input parameters, not based on some other precomputed data that I then need to understand and match with this logic.